prow/cmd/branchprotector/protect.go

```diff
@@ -315,67 +306,95 @@ func (p *protector) UpdateBranch(orgName, repo string, branchName string, branch
+	default:
+		return false
+	}
+}
+
+func equalRequiredStatusChecks(state, request *github.RequiredStatusChecks) bool {
+	switch {
+	case state == request:
+		return true
+	case state != nil && request != nil:
+		return state.Strict == request.Strict &&
+			equalStringSlices(&state.Contexts, &request.Contexts)
+	default:
+		return false
+	}
+}
+
+func equalStringSlices(s1, s2 *[]string) bool {
+	switch {
+	case s1 == s2:
+		return true
+	case s1 != nil && s2 != nil:
+		if len(*s1) != len(*s2) {
+			return false
+		}
+		sort.Strings(*s1)
+		sort.Strings(*s2)
+		for i, v := range *s1 {
+			if v != (*s2)[i] {
 				return false
 			}
-			sort.Strings(*s1)
-			sort.Strings(*s2)
-			for i, v := range *s1 {
-				if v != (*s2)[i] {
-					return false
-				}
-			}
-			return true
-		default:
-			return false
 		}
+		return true
+	default:
+		return false
 	}
+}
 
-	equalStatusChecks := func(sc1, sc2 *github.RequiredStatusChecks) bool {
-		switch {
-		case sc1 == sc2:
-			return true
-		case sc1 != nil && sc2 != nil:
-			return sc1.Strict == sc2.Strict &&
-				equalStringSlices(&sc1.Contexts, &sc2.Contexts)
-		default:
-			return false
-		}
+func equalAdminEnforcement(state github.EnforceAdmins, request *bool) bool {
+	switch {
+	case request == nil:
+		// requesting nothing is the same as requesting this to be off
+		return state.Enabled == false
+	default:
+		return state.Enabled == *request
 	}
+}
 
-	equalRestrictions := func(r1, r2 *github.Restrictions) bool {
-		switch {
-		case r1 == r2:
-			return true
-		case r1 != nil && r2 != nil:
-			return equalStringSlices(r1.Teams, r2.Teams) && equalStringSlices(r1.Users, r2.Users)
-		default:
-			return false
-		}
+func equalRequiredPullRequestReviews(state *github.RequiredPullRequestReviews, request *github.RequiredPullRequestReviewsRequest) bool {
+	switch {
+	case state == nil && request == nil:
+		return true
+	case state != nil && request != nil:
+		return state.DismissStaleReviews == request.DismissStaleReviews &&
+			state.RequireCodeOwnerReviews == request.RequireCodeOwnerReviews &&
+			state.RequiredApprovingReviewCount == request.RequiredApprovingReviewCount &&
+			equalRestrictions(state.DismissalRestrictions, &request.DismissalRestrictions)
+	default:
+		return false
 	}
+}
 
-	equalReviews := func(r1, r2 *github.RequiredPullRequestReviews) bool {
-		switch {
-		case r1 == r2:
-			return true
-		case r1 != nil && r2 != nil:
-			return r1.DismissStaleReviews == r2.DismissStaleReviews &&
-				r1.RequireCodeOwnerReviews == r2.RequireCodeOwnerReviews &&
-				r1.RequiredApprovingReviewCount == r2.RequiredApprovingReviewCount &&
-				equalRestrictions(&r1.DismissalRestrictions, &r2.DismissalRestrictions)
-		default:
-			return false
+func equalRestrictions(state *github.Restrictions, request *github.RestrictionsRequest) bool {
+	switch {
+	case state == nil && request == nil:
+		return true
+	case state == nil && request != nil:
+		return request.Users == nil && request.Teams == nil
```

Comment:
It's just definition -- but yes

Reply:
Technically here we are actually setting two things to be equal that are not equal. However, I don't think it matters?

The setting in question is this:
![Screenshot from 2019-06-08 08-17-29](https://user-images.githubusercontent.com/7328370/59149103-25aaef00-89c6-11e9-979a-030809fbec9f.png)


I have explicitly checked to enable restrictions, but I have also not actually committed any. I checked, and doing this will persist this weird state, it won't revert to having the check-box un-checked. So if we care, we could call that state unequal to:

![Screenshot from 2019-06-08 08-20-41](https://user-images.githubusercontent.com/7328370/59149112-51c67000-89c6-11e9-8e1e-2e9a7e2417f5.png)


However from a functionality standpoint it's moot.